machine/v1alpha1/provider_types.go

```diff
@@ -0,0 +1,207 @@
+
+	// Verbosity set the logs verbosity. Defaults to 1.
+	// Controller Manager flag is --verbosity.
+	// +optional
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=0
+	Verbosity int `json:"verbosity,omitempty"`
+
+	// Debug, if set, will override a set of fields with opinionated values for
+	// a debugging session. (Verbosity=5, ProfilerAddress=localhost:6060)
+	// +optional
+	// +kubebuilder:default=false
+	Debug bool `json:"debug,omitempty"`
+
+	// FeatureGates define provider specific feature flags that will be passed
+	// in as container args to the provider's controller manager.
+	// Controller Manager flag is --feature-gates.
+	FeatureGates map[string]bool `json:"featureGates,omitempty"`
+}
+
+// DeploymentSpec defines the properties that can be enabled on the Deployment for the provider.
+type DeploymentSpec struct {
+	// Number of desired pods. This is a pointer to distinguish between explicit zero and not specified. Defaults to 1.
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	Replicas *int `json:"replicas,omitempty"`
+
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// If specified, the pod's tolerations.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// If specified, the pod's scheduling constraints
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty"`
+
+	// List of containers specified in the Deployment
+	// +optional
+	Containers []ContainerSpec `json:"containers"`
+}
+
+// ContainerSpec defines the properties available to override for each
+// container in a provider deployment such as Image and Args to the container’s
+// entrypoint.
+type ContainerSpec struct {
+	// Name of the container. Cannot be updated.
+	Name string `json:"name"`
+
+	// Container Image Name
+	// +optional
+	Image *ImageMeta `json:"image,omitempty"`
+
+	// Args represents extra provider specific flags that are not encoded as fields in this API.
+	// Explicit controller manager properties defined in the `Provider.ManagerSpec`
+	// will have higher precedence than those defined in `ContainerSpec.Args`.
+	// For example, `ManagerSpec.SyncPeriod` will be used instead of the
+	// container arg `--sync-period` if both are defined.
+	// The same holds for `ManagerSpec.FeatureGates` and `--feature-gates`.
+	// +optional
+	Args map[string]string `json:"args,omitempty"`
+
+	// List of environment variables to set in the container.
+	// +optional
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Compute resources required by this container.
+	// +optional
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// ImageMeta allows to customize the image used.
+type ImageMeta struct {
```

Comment:
Do the fields within here need to be pointers? Are empty strings valid values?

Reply:
They don't need to be pointers and empty strings are valid.